extensions/sql/processors/QueryDatabaseTable.cpp

```diff
@@ -0,0 +1,465 @@
+    mapState_.clear();
+  } else {
+    for (const auto& columName : listMaxValueColumnName_) {
+      if (0 == mapState_.count(columName)) {
+        mapState_.clear();
+        break;
+      }
+    }
+  }
+
+  // Add in 'mapState_' new columns which are in 'listMaxValueColumnName_'.
+  for (const auto& maxValueColumnName: listMaxValueColumnName_) {
+    if (0 == mapState_.count(maxValueColumnName)) {
+      mapState_.insert({maxValueColumnName, std::string()});
+    }
+  }
+
+  const auto dynamic_prop_keys = context->getDynamicPropertyKeys();
+  logger_->log_info("Received %zu dynamic properties", dynamic_prop_keys.size());
+
+  // If the stored state for a max value column is empty, populate it with the corresponding initial max value, if it exists.
+  for (const auto& key : dynamic_prop_keys) {
+    if (std::string::npos == key.rfind(s_initialMaxValueDynamicPropertyPrefix, 0)) {
+      throw minifi::Exception(PROCESSOR_EXCEPTION, "QueryDatabaseTable: Unsupported dynamic property \"" + key + "\"");
+    }
+    const auto columnName = utils::toLower(key.substr(s_initialMaxValueDynamicPropertyPrefix.length()));
+    auto it = mapState_.find(columnName);
+    if (it == mapState_.end()) {
+      logger_->log_warn("Initial maximum value specified for column \"%s\", which is not specified as a Maximum-value Column. Ignoring.", columnName);
+      continue;
+    }
+    if (!it->second.empty()) {
+      continue;
+    }
+    std::string value;
+    if (context->getDynamicProperty(key, value) && !value.empty()) {
+      it->second = value;
+      logger_->log_info("Setting initial maximum value of %s to %s", columnName, value);
+    }
+  }
+}
+
+void QueryDatabaseTable::processOnTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
+  const auto& selectQuery = getSelectQuery();
+
+  logger_->log_info("QueryDatabaseTable: selectQuery: '%s'", selectQuery.c_str());
+
+  auto statement = connection_->prepareStatement(selectQuery);
+
+  auto rowset = statement->execute();
+
+  int count = 0;
+  size_t rowCount = 0;
+  sql::MaxCollector maxCollector(selectQuery, maxValueColumnNames_, mapState_);
+  sql::JSONSQLWriter jsonSQLWriter;
+  sql::SQLRowsetProcessor sqlRowsetProcessor(rowset, {&jsonSQLWriter, &maxCollector});
+
+  // Process rowset.
+  do {
+    rowCount = sqlRowsetProcessor.process(maxRowsPerFlowFile_ == 0 ? std::numeric_limits<size_t>::max() : maxRowsPerFlowFile_);
+    count++;
+    if (rowCount == 0)
+      break;
+
+    const auto& output = jsonSQLWriter.toString();
+    if (!output.empty()) {
+      WriteCallback writer(output.data(), output.size());
+      auto newflow = session->create();
+      newflow->addAttribute(ResultRowCount, std::to_string(rowCount));
+      newflow->addAttribute(ResultTableName, tableName_);
+      session->write(newflow, &writer);
+      session->transfer(newflow, s_success);
+    }
+  } while (rowCount > 0);
+
+  if (maxCollector.updateMapState()) {
```

Comment:
We should only call `updateMapState`, and thereby modify the in-memory state AFTER a successful `session->commit()`, otherwise we risk data loss.

Reply:
Fixed.